### DISCUSSION TITLE
Update getCollection to return a copy instead of the same reference

### DISCUSSION THREAD

**Discussion by kevinqdam** [withastro/roadmap#482] [Proposal]
### Body

# Summary
The purpose of this RFC is to propose that `getCollection` from `"astro:content"` return a copy of the collection for immutability instead of the same reference across multiple calls to `getCollection`.

Currently, `getCollection` returns the content entries by reference. Please see this [line](https://github.com/withastro/astro/blob/8b7cb64dadfca93c65d62df54754633d398cb2ed/packages/astro/src/content/internal.ts#L50). Significantly, the `entry.data` is the same reference which can have unwanted side effects across pages, even when changing only one page file.

# Background & Motivation

I ran into this when writing my `/blog` and `/blog/tag/[tag]` pages.
- `/blog` should show all posts and the tags that the reader can browse by.
- `/blog/tag/[tag]` shows all posts that have the tag in the path. It is the same as `/blog` except with a filtered list of posts and styling to indicate the current tag.

Please check out the examples of both below when working correctly and when encountering the side effects from mutation:

## Copying the content collection works as expected
**Implementation (copy the content collection, don't mutate it)**
```ts
// pages/blog/tag/[tag].astro

const tagSchema = z.object({
  text: z.enum([ 'web-development', 'astro' ]),
  href: z.string(),
});

type Tag = z.infer<typeof tagSchema>;

type Pill = Tag & {
  invertColor?: boolean;
};

const { tag: currentTag } = Astro.params;

const getUniqueTags = (
  tags: CollectionEntry<'blog'>['data']['tags']
): Pill[] => {
  const result: Pill[] = [];
  const seen = new Set();
  for (const tag of tags) {
    if (seen.has(tag.text)) {
      continue;
    }
    seen.add(tag.text);
    // copy the tag before styling it as an inverted Pill
    result.push(
      tag.text === currentTag
        ? { ...tag, invertColor: true, href: '/blog' }
        : { ...tag }
    );
  }
  return result;
};
```

**`/blog` is not styled with inverted tags as expected**
<img width="1290" alt="blog" src="https://user-images.githubusercontent.com/43325829/218283811-d0f74bcf-b674-4bd2-a2d7-ed488bdb603a.png">

**`/blog/tag/[tag]` is styled with the correct inverted tag as expected**
<img width="1294" alt="blog-tag" src="https://user-images.githubusercontent.com/43325829/218283819-b5dfe84a-8806-4b71-b177-3d77eddf0dcc.png">

## Mutating the content collection will affect other pages
**Implementation (mutate the content collection object)**
```ts
// pages/blog/tag/[tag].astro

const tagSchema = z.object({
  text: z.enum([ 'web-development', 'astro' ]),
  href: z.string(),
});

type Tag = z.infer<typeof tagSchema>;

type Pill = Tag & {
  invertColor?: boolean;
};

const { tag: currentTag } = Astro.params;

const getUniqueTags = (
  tags: CollectionEntry<'blog'>['data']['tags']
): Pill[] => {
  const result: Pill[] = [];
  const seen = new Set();
  for (const tag of tags) {
    if (seen.has(tag.text)) {
      continue;
    }
    seen.add(tag.text);
    // mutate the tag if it is the current tag so it is styled correctly
    if (tag.text === currentTag) {
      (tag as Pill).invertColor = true;
      (tag as Pill).href = '/blog';
    }
    result.push(tag);
  }
  return result;
};
```

**`/blog` styling is affected, even when I only changed `/blog/tag/[tag]`**
<img width="1293" alt="blog2" src="https://user-images.githubusercontent.com/43325829/218284625-4fdfea58-ce6b-4ec5-a5eb-ee4179d62c87.png">

**`/blog/tag/[tag]` doesn't look too good either, tags have inverted colors and it is not clear why**
<img width="1292" alt="blog-tag2" src="https://user-images.githubusercontent.com/43325829/218284626-73221f76-2096-49a8-87bf-1f589e58cb14.png">

# Goals
- The goal is to abide by the principle of least surprise. Specifically, we might say that mutating the collection on the current page file should only affect the current page. For example, making a change in `pages/blog/foo` and seeing it affect `pages/blog/bar` is very surprising.
- Another goal is to assess the tradeoff. Copying the collection object is potentially expensive.
    - That is, building the Astro site like this can take longer if we copy the collection object every time. This cost might be worth it for a more predictable `getCollection` API.
    - If build time is more valuable, then we might leave it up to the developer to copy the content collection object whenever they want to do something in a mutable way like the failure mode I showed above.

# Example
The proposed change is to change [this line](https://github.com/withastro/astro/blob/8b7cb64dadfca93c65d62df54754633d398cb2ed/packages/astro/src/content/internal.ts#L50).

```ts
// Before
const entry = await lazyImport();

// After (just one potential approach to get the point across)
const entryCopy = JSON.parse(JSON.stringify(await lazyImport()));
```

The above example is just one way of creating a copy. I tried it out locally, and the cross-page side effects went away. Maybe `lodash`'s [`_.cloneDeep`](https://lodash.com/docs/#cloneDeep) might be another option here, if we are keen on changing this and don't like the stringify-then-parse approach?

# Conclusion and thank you
I thought this was worth raising for newer programmers who aren't keen on the dangers of mutation and/or the differences between values and references in JavaScript.

Thank you so much for your consideration and all of the hard work done for Astro! I am enjoying Astro a ton and am blown away by how easy and fast it is to get going. Great work everybody, I am rooting for you!!

**Comment by florian-lefebvre** [comment]
This should now be the case!
